test/unit/org/apache/cassandra/cql3/DurationTest.java

```diff
@@ -174,7 +175,211 @@ private long toMillis(String timeAsString)
         return DateUtils.truncate(date, Calendar.SECOND).getTime();
     }
 
-    public void assertInvalidDuration(String duration, String expectedErrorMessage)
+    @Test
+    public void testInvalidFloorTimestamp()
+    {
+        try
+        {
+            floorTimestamp("2016-09-27T16:12:00", "2h", "2017-09-01T00:00:00");
+            Assert.fail();
+        }
+        catch (InvalidRequestException e)
+        {
+            assertEquals("The floor function starting time is greater than the provided time", e.getMessage());
+        }
+
+        try
+        {
+            floorTimestamp("2016-09-27T16:12:00", "-2h", "2016-09-27T00:00:00");
+            Assert.fail();
+        }
+        catch (InvalidRequestException e)
+        {
+            assertEquals("Negative durations are not supported by the floor function", e.getMessage());
+        }
+    }
+
+    @Test
+    public void testFloorTimestamp()
+    {
+        //~ Test with a duration in hours //////////////////////////////////////////////
+
+        // Test floor with a timestamp equals to the start time
+        long result = floorTimestamp("2016-09-27T16:12:00", "2h", "2016-09-27T16:12:00");
+        assertEquals(toMillis("2016-09-27T16:12:00"), result);
```

Comment:
I found it is easier to read the assertion error in the form of human-readable format.
for example, this is hard to correlate with the test. 
```
java.lang.AssertionError: 
Expected :1473469200000
Actual   :1473494400000
```

So instead of `toMillis`, how about convert a millis value to a date time string and just compare the string value?